// @relaypro/sdk/dist/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Notification = exports.DeviceInfoField = exports.DeviceInfoQuery = exports.Language = exports.Taps = exports.Button = exports.Event = void 0;
var Event;
(function (Event) {
    Event["START"] = "start";
    Event["BUTTON"] = "button";
    Event["TIMER"] = "timer";
    Event["NOTIFICATION"] = "notification";
    Event["CALL_CONNECTED"] = "call_connected";
    Event["CALL_DISCONNECTED"] = "call_disconnected";
    Event["CALL_FAILED"] = "call_failed";
    Event["CALL_RECEIVED"] = "call_received";
    Event["CALL_START_REQUEST"] = "call_start_request";
})(Event = exports.Event || (exports.Event = {}));
var Button;
(function (Button) {
    Button["ACTION"] = "action";
    Button["CHANNEL"] = "channel";
})(Button = exports.Button || (exports.Button = {}));
var Taps;
(function (Taps) {
    Taps["SINGLE"] = "single";
    Taps["DOUBLE"] = "double";
    Taps["TRIPLE"] = "triple";
})(Taps = exports.Taps || (exports.Taps = {}));
var Language;
(function (Language) {
    Language["ENGLISH"] = "en-US";
    Language["GERMAN"] = "de-DE";
    Language["SPANISH"] = "es-ES";
    Language["FRENCH"] = "fr-FR";
    Language["ITALIAN"] = "it-IT";
    Language["RUSSIAN"] = "ru-RU";
    Language["SWEDISH"] = "sv-SE";
    Language["TURKISH"] = "tr-TR";
    Language["HINDI"] = "hi-IN";
    Language["ICELANDIC"] = "is-IS";
    Language["JAPANESE"] = "ja-JP";
    Language["KOREAN"] = "ko-KR";
    Language["POLISH"] = "ko-KR";
    Language["PORTUGUESE"] = "pt-BR";
    Language["NORWEIGN"] = "nb-NO";
    Language["DUTCH"] = "nl-NL";
    Language["CHINESE"] = "zh";
})(Language = exports.Language || (exports.Language = {}));
var DeviceInfoQuery;
(function (DeviceInfoQuery) {
    DeviceInfoQuery["NAME"] = "name";
    DeviceInfoQuery["ID"] = "id";
    DeviceInfoQuery["ADDRESS"] = "address";
    DeviceInfoQuery["COORDINATES"] = "latlong";
    DeviceInfoQuery["BATTERY"] = "battery";
    DeviceInfoQuery["INDOOR_LOCATION"] = "indoor_location";
    DeviceInfoQuery["LOCATION"] = "location";
})(DeviceInfoQuery = exports.DeviceInfoQuery || (exports.DeviceInfoQuery = {}));
var DeviceInfoField;
(function (DeviceInfoField) {
    DeviceInfoField["LABEL"] = "label";
    DeviceInfoField["CHANNEL"] = "channel";
})(DeviceInfoField = exports.DeviceInfoField || (exports.DeviceInfoField = {}));
var Notification;
(function (Notification) {
    Notification["BROADCAST"] = "broadcast";
    Notification["ALERT"] = "alert";
    Notification["NOTIFY"] = "notify";
    Notification["CANCEL"] = "cancel";
})(Notification = exports.Notification || (exports.Notification = {}));
