// @relaypro/sdk/dist/queue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Queue {
    constructor(params = {}) {
        Object.defineProperty(this, "max", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "numActive", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "queue", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        const { max = 1 } = params;
        this.max = max;
        this.numActive = 0;
        this.queue = [];
    }
    next() {
        if (this.queue.length) {
            if (this.numActive < this.max) {
                const fn = this.queue.shift();
                fn && this.execute(fn);
            }
        }
    }
    execute(fn) {
        this.numActive++;
        new Promise(resolve => {
            const obj = fn();
            if (typeof obj?.then === `function`) {
                resolve();
            }
        }).finally(() => {
            this.numActive--;
            this.next();
        });
    }
    enqueue(fn) {
        this.queue.push(fn);
        this.next();
    }
}
exports.default = Queue;
